### PR TITLE
Implement Space function push down #5113

### LINKED ISSUE #5113: Implement `Space` function push down
## Feature Request

**Describe the feature you'd like:**

We'd like to support `Space` function push down.

**Teachability, Documentation, Adoption, Migration Strategy:**

At least one integration test and several unit tests.

### PR DESCRIPTION
**Pull request by lizhenhuan**
### What problem does this PR solve?

Issue Number: close #5113

Problem Summary: Implement Space function push down #5113
Related PR:https://github.com/pingcap/tidb/pull/35753

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support to pushdown space to TiFlash
```

### CODE REVIEW

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+protected:
+    ColumnWithTypeAndName toVec(const std::vector<std::optional<String>> & v)
+    {
+        return createColumn<Nullable<String>>(v);
+    }
+};
+
+// test string and string
+TEST_F(StringReverse, strAndStrTest)
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:51]: StringSpace?

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    ASSERT_COLUMN_EQ(
+        toVec({"  ", "", "          "}),
+        executeFunction(
+            func_name,
+            toVec({2, 0, 10})));
+}
+CATCH
+
+// test NULL
+TEST_F(StringLength, nullTest)
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:63]: ditto

File: dbms/src/Functions/FunctionsString.cpp [hunk 3]
```diff
@@ -4188,6 +4188,54 @@ class FunctionLength : public IFunction
...
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        const ColumnPtr c0_col = block.getByPosition(arguments[0]).column;
+
+        Field res_field;
+        const auto * column_space_num = checkAndGetColumn<ColumnConst>(c0_col.get());
+        Int64 space_num = 0;
+
+        if (column_space_num)
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsString.cpp:4224]:
```suggestion
        if (!column_space_num)
```
Maybe this is your intention?

File: dbms/src/Functions/FunctionsString.cpp [hunk 4]
```diff
@@ -4188,6 +4188,54 @@ class FunctionLength : public IFunction
...
+    {
+        return std::make_shared<FunctionSpace>();
+    }
+
+    std::string getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 1)
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4208]: Don't need to check argument number here if [isVariadic()](https://github.com/pingcap/tiflash/blob/0df17c6ea5a1d7f18fe1709916dbfa98b7ed24ef/dbms/src/Functions/IFunction.h#L192) is false

File: dbms/src/Functions/FunctionsString.cpp [hunk 5]
```diff
@@ -4188,6 +4188,54 @@ class FunctionLength : public IFunction
...
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        const ColumnPtr c0_col = block.getByPosition(arguments[0]).column;
+
+        Field res_field;
+        const auto * column_space_num = checkAndGetColumn<ColumnConst>(c0_col.get());
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4221]: Seems you assuming the input column is always a constant? Please check if `TiDB/MySQL` support non-constant column for function `space`, if yes, then I suggest:
- Support non-constant column as well in TiFlash
- Leave a todo here to support non constant column in the furture

File: dbms/src/Functions/FunctionsString.cpp [hunk 6]
```diff
@@ -4188,6 +4188,54 @@ class FunctionLength : public IFunction
...
+
+        if (!column_space_num)
+        {
+            space_num = column_space_num->getInt(0);
+        }
+        auto col_res = ColumnString::create();
+        col_res->reserve(space_num);
+
+        std::string res_string(space_num, ' ');
+        col_res->insert(res_string);
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4232]: I'm a little confused that why you only insert one row here? The row number of `col_res` should be the same as the row number of the input `block`.

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+{
+class StringSpace : public DB::tests::FunctionTest
+{
+public:
+    static constexpr auto func_name = "space";
+
+protected:
+    ColumnWithTypeAndName toVec(const std::vector<std::optional<String>> & v)
+    {
+        return createColumn<Nullable<String>>(v);
```
Comment (windtalker) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:46]: Since TiDB will ensure that the input type of `space` will be integer, it's meaningless to use `String` column as the input.

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionsString.h>
+#include <Interpreters/Context.h>
+#include <TestUtils/FunctionTestUtils.h>
+#include <TestUtils/TiFlashTestBasic.h>
+
+#include <string>
+#include <vector>
+
+#pragma GCC diagnostic pop
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:28]: Why use this #pragma?

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionsString.h>
+#include <Interpreters/Context.h>
+#include <TestUtils/FunctionTestUtils.h>
+#include <TestUtils/TiFlashTestBasic.h>
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:23]: Some of them are useless, you can remove them.

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+// test space
+TEST_F(StringSpace, spaceTest)
+try
+{
+    ASSERT_COLUMN_EQ(
+        toVec({"  ", "", "          "}),
+        executeFunction(
+            func_name,
+            toVecInt({2, 0, 10})));
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:56]: if the input field is -1 or bigger than Int64_MAX, what will happen?

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+// test space NULL
+TEST_F(StringSpace, nullTest)
+try
+{
+    ASSERT_COLUMN_EQ(
+        toVec({{}}),
+        executeFunction(
+            func_name,
+            toVec({{}})));
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:68]: Use toVecInt instead of toVec maybe better.

File: tests/fullstack-test/expr/space.test [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+mysql> drop table if exists test.t;
+mysql> create table if not exists test.t(a varchar(20));
+
+mysql> insert into test.t values('www.pingcap.com');
```
Comment (ywqzzy) [inline tests/fullstack-test/expr/space.test:18]: You can create table t with Int64 columns, then call space(t.xx) to test.

File: dbms/src/Functions/FunctionsString.cpp [hunk 13]
```diff
@@ -4396,6 +4396,54 @@ class FunctionRepeat : public IFunction
...
+
+        Field res_field;
+        int val_num = c0_col->size();
+        auto col_res = ColumnString::create();
+        col_res->reserve(val_num);
+
+        for (int i = 0; i < val_num; i++)
+        {
+            c0_col->get(i, res_field);
+            Int64 space_num = res_field.get<Int64>();
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4436]: need handle negative value here.

Comment (windtalker) [reply dbms/src/Functions/FunctionsString.cpp:4436]: And according to MySQL's implementation, the max value should be Int32_max

File: dbms/src/Functions/FunctionsString.cpp [hunk 14]
```diff
@@ -4396,6 +4396,54 @@ class FunctionRepeat : public IFunction
...
+        const ColumnPtr c0_col = block.getByPosition(arguments[0]).column;
+
+        Field res_field;
+        int val_num = c0_col->size();
+        auto col_res = ColumnString::create();
+        col_res->reserve(val_num);
+
+        for (int i = 0; i < val_num; i++)
+        {
+            c0_col->get(i, res_field);
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4435]: Don't need to get res_field everytime, it's in-efficient, you can refer to `TidbPadImpl` as an example.

Comment (windtalker) [reply dbms/src/Functions/FunctionsString.cpp:4435]: Also need to some optimzations for constant column.

File: dbms/src/Functions/tests/gtest_strings_space.cpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+{
+    ASSERT_COLUMN_EQ(
+        toNullableVec({"  ", "", "          ", "", "      "}),
+        executeFunction(
+            func_name,
+            toVecInt({2, 0, 10, -1, 6})));
+}
+CATCH
+
+// test space NULL
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_space.cpp:65]: maybe add more test with null and not null？

**Review by ywqzzy — COMMENTED** [review]
Others LGTM

File: dbms/src/Functions/FunctionsString.cpp [hunk 16]
```diff
@@ -4392,6 +4392,109 @@ class FunctionRepeat : public IFunction
 };
 
 
+class FunctionSpace : public IFunction
+{
+public:
+    static constexpr auto name = "space";
+
+    static constexpr auto MAX = 16777216;
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsString.cpp:4400]: Why use this magic number？

File: dbms/src/Functions/FunctionsString.cpp [hunk 17]
```diff
@@ -4392,6 +4392,109 @@ class FunctionRepeat : public IFunction
...
+    }
+
+    std::string getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        return arguments[0]->onlyNull()
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4416]: Usually, we also check the argument type is valid or not in `getReturnTypeImpl`

File: dbms/src/Functions/FunctionsString.cpp [hunk 18]
```diff
@@ -4392,6 +4392,109 @@ class FunctionRepeat : public IFunction
...
+
+        col_res_offsets.resize(c0_col_column->size());
+        ColumnString::Offset res_offset = 0;
+
+
+        for (size_t row = 0; row < val_num; ++row)
+        {
+            result_null_map->getData()[row] = false;
+
+            Int64 space_num = accurate::lessOp(INT64_MAX, col_vector_space_num_value[row]) ? INT32_MAX : col_vector_space_num_value[row];
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4472]: Why compare with `INT64_MAX` and use `INT32_MAX` latter?

File: dbms/src/Functions/FunctionsString.cpp [hunk 19]
```diff
@@ -4392,6 +4392,109 @@ class FunctionRepeat : public IFunction
...
+            if (space_num < 0)
+            {
+                space_num = 0;
+            }
+            if (space_num > MAX)
+            {
+                result_null_map->getData()[row] = true;
+                space_num = 0;
+            }
+            col_res_data.resize(col_res_data.size() + space_num + 1);
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4482]: better to reserve some space before the loop.

Comment (windtalker) [reply dbms/src/Functions/FunctionsString.cpp:4482]: And also since in most of the user case, the argument of `space` should be a constant, so I think we should not set  `useDefaultImplementationForConstants` to true, and handle constant column explicitly.

File: dbms/src/Functions/FunctionsString.cpp (hunk 16 above)
Comment (lizhenhuan) [reply dbms/src/Functions/FunctionsString.cpp:4400]: Because when input bigger than 16777216(32 power of 2) then the space function will return NULL
image

File: dbms/src/Functions/FunctionsString.cpp [hunk 20]
```diff
@@ -4392,6 +4392,142 @@ class FunctionRepeat : public IFunction
...
+            throw Exception(
+                fmt::format("Illegal type {} of first argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        return arguments[0]->onlyNull()
+            ? makeNullable(std::make_shared<DataTypeNothing>())
+            : makeNullable(std::make_shared<DataTypeString>());
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsString.cpp:4425]: You can check return type using getReturnTypeImpl here

File: dbms/src/Functions/FunctionsString.cpp [hunk 21]
```diff
@@ -4392,6 +4392,142 @@ class FunctionRepeat : public IFunction
...
+
+        size_t val_num = block.rows();
+        auto result_null_map = ColumnUInt8::create(val_num);
+        auto col_res = ColumnString::create();
+        auto & col_res_data = col_res->getChars();
+        auto & col_res_offsets = col_res->getOffsets();
+        col_res_data.reserve(c0_col_column->size());
+        col_res_offsets.resize(c0_col_column->size());
+        ColumnString::Offset res_offset = 0;
+        if (c0_col_column->isColumnConst())
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsString.cpp:4463]: extract a executeConst function and a executeVector function

**Review by ywqzzy — APPROVED** [review]
Others LGTM

File: dbms/src/Functions/FunctionsString.cpp [hunk 22]
```diff
@@ -4392,6 +4392,168 @@ class FunctionRepeat : public IFunction
...
+        auto & col_res_data = col_res->getChars();
+        auto & col_res_offsets = col_res->getOffsets();
+        col_res_data.reserve(c0_col_column->size());
+        col_res_offsets.resize(c0_col_column->size());
+
+
+        if (c0_col_column->isColumnConst())
+        {
+            const ColumnConst * col_const_space_num = checkAndGetColumnConst<ColumnVector<IntType>>(c0_col_column.get());
+            auto space_num_values = col_const_space_num->getValue<IntType>();
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4467]: need to check if `col_const_space_num` is nullptr or not

File: dbms/src/Functions/FunctionsString.cpp [hunk 23]
```diff
@@ -4392,6 +4392,168 @@ class FunctionRepeat : public IFunction
...
+    }
+
+    static void executeConst(
+        Int64 space_num,
+        size_t val_num,
+        ColumnUInt8::Container & result_null_map_data,
+        ColumnString::Chars_t & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        ColumnString::Offset res_offset = 0;
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4493]: Should reserve first

File: dbms/src/Functions/FunctionsString.cpp [hunk 24]
```diff
@@ -4392,6 +4392,174 @@ class FunctionRepeat : public IFunction
...
+        auto c0_col_column = c0_col.column;
+
+        size_t val_num = block.rows();
+        auto result_null_map = ColumnUInt8::create(val_num);
+        auto col_res = ColumnString::create();
+        auto & col_res_data = col_res->getChars();
+        auto & col_res_offsets = col_res->getOffsets();
+
+        col_res_offsets.reserve(c0_col_column->size());
+        col_res_offsets.resize(c0_col_column->size());
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4463]: just call resize is enough

File: dbms/src/Functions/FunctionsString.cpp [hunk 25]
```diff
@@ -4392,6 +4392,174 @@ class FunctionRepeat : public IFunction
...
+
+    static void executeConst(
+        Int64 space_num,
+        size_t val_num,
+        ColumnUInt8::Container & result_null_map_data,
+        ColumnString::Chars_t & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        ColumnString::Offset res_offset = 0;
+        res_data.reserve(val_num * (space_num + 1));
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4499]: need to check `space_num > MAX_BLOB_WIDTH` before reserve. If `space_num > MAX_BLOB_WIDTH` is true, just reserve `val_num` is enough.

File: dbms/src/Functions/FunctionsString.cpp [hunk 26]
```diff
@@ -4392,6 +4392,183 @@ class FunctionRepeat : public IFunction
...
+        if (space_num > MAX_BLOB_WIDTH)
+        {
+            res_data.reserve(val_num);
+            is_big = true;
+        }
+        else
+        {
+            res_data.reserve(val_num * (space_num + 1));
+        }
+        if (space_num < 0)
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4509]: This check should be moved before reserve

File: dbms/src/Functions/FunctionsString.cpp [hunk 27]
```diff
@@ -4392,6 +4392,183 @@ class FunctionRepeat : public IFunction
...
+        ColumnString::Chars_t & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        ColumnString::Offset res_offset = 0;
+        auto is_big = false;
+
+        if (space_num > MAX_BLOB_WIDTH)
+        {
+            res_data.reserve(val_num);
+            is_big = true;
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4503]: why not just set `space_num` to 0 here

File: dbms/src/Functions/FunctionsString.cpp [hunk 28]
```diff
@@ -4392,6 +4392,183 @@ class FunctionRepeat : public IFunction
...
+            result_null_map_data[row] = false;
+
+            if (is_big)
+            {
+                result_null_map_data[row] = true;
+                space_num = 0;
+            }
+            res_data.resize(res_data.size() + space_num + 1);
+
+            std::string res_string(space_num, ' ');
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4525]: this can be move before the for loop.

File: dbms/src/Functions/FunctionsString.cpp [hunk 29]
```diff
@@ -4392,6 +4392,184 @@ class FunctionRepeat : public IFunction
...
+private:
+    template <typename IntType>
+    bool executeSpace(
+        Block & block,
+        const ColumnNumbers & arguments,
+        const size_t result) const
+    {
+        auto & c0_col = block.getByPosition(arguments[0]);
+
+        Field res_field;
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4453]: useless variable?

**Review by windtalker — APPROVED** [review]
Other LGTM

File: dbms/src/Functions/FunctionsString.cpp (hunk 29 above)
Comment (lizhenhuan) [reply dbms/src/Functions/FunctionsString.cpp]: deleted
